Create component tutorial (#208)
Financial-Times/origami-website: _tutorials/new-component/create-a-new-component-part-1.md

---
title: Create A New Origami Component
description: A step-by-step tutorial which teaches you how to build and deploy a new Origami component.
cta: Learn how to create an Origami component

---

# {{page.title}}

The "Create A New Origami Component" tutorial is split into eight parts and is intended to be followed sequentially from start to finish:

> - part one: Intro & Boilerplate
> - [part two](/docs/tutorials/create-a-new-component-part-2/): Base Styles
> - [part three](/docs/tutorials/create-a-new-component-part-3/): Themes & Brands
> - [part four](/docs/tutorials/create-a-new-component-part-4/): Demos
> - [part five](/docs/tutorials/create-a-new-component-part-5/): JavaScript
> - [part six](/docs/tutorials/create-a-new-component-part-6/): Testing
> - [part seven](/docs/tutorials/create-a-new-component-part-7/): Documentation
> - [part eight](/docs/tutorials/create-a-new-component-part-8/): Component Lifecycle
> 

Reviewer: What do we think to this being an ordered list?
Before:
<img width="555" alt="image" src="https://user-images.githubusercontent.com/1569131/87307281-33a83900-c511-11ea-9b96-d52458d9138b.png">

After:
<img width="626" alt="image" src="https://user-images.githubusercontent.com/1569131/87307261-28eda400-c511-11ea-96df-3a56d796a260.png">

```suggestion
-

1. Intro & Boilerplate
2. [Base Styles](/docs/tutorials/create-a-new-component-part-2/)
3. [Themes & Brands](/docs/tutorials/create-a-new-component-part-3/)
4. [Demos](/docs/tutorials/create-a-new-component-part-4/)
5. [JavaScript](/docs/tutorials/create-a-new-component-part-5/)
6. [Testing](/docs/tutorials/create-a-new-component-part-6/)
7. [Documentation](/docs/tutorials/create-a-new-component-part-7/)
8. [Component Lifecycle](/docs/tutorials/create-a-new-component-part-8/)

```

## Introduction

Usually the Origami team will own an Origami component which is used by multiple teams or groups, but anybody can create and share an Origami component.

> In this tutorial we'll build an Origami component. Our example component will display a "hello world" message and include a button which will count the number of times it was clicked. Visually the component will include multiple variations to suit reader facing and internal projects. We'll also discuss how to document and publish an Origami component so it is visible in the [Origami Registry](https://registry.origami.ft.com/components).

Reviewer: >reader facing and internal projects
I've never seen the term "reader facing" before, do we mean master brand? If we mean master brand and we think "reader facing" is useful to those who don't know what master brand means, we could put master brand in brackets so that the user knows to relate those terms together.

>reader facing (master brand) and internal projects (internal brand).

> ## Prerequisites

Reviewer: This needs a section where we mention the software required to follow the tutorial.

They will need:

- NodeJS (Version 10 or higher)
- npm (Version 5 or higher)
- Bower

If you have any questions please contact the Origami team to help make this tutorial better &#x1F603;. You can find the team on Slack in <a href="https://financialtimes.slack.com/messages/{{site.data.contact.slack}}" class="o-typography-link--external" target="_blank">#{{site.data.contact.slack}}</a>.

## The Origami Specification

All Origami components **must** meet the requirements defined in the [Origami Component Specification](/spec/v1/components/). The specification covers folder structure, code standards, documentation and more. We'll refer to this document throughout this tutorial.

<aside>
If this tutorial contradicts or misses anything required in the <a href="/spec/v1/components/">Origami Component Specification</a>, the specification is correct and this tutorial is wrong. If you find a case like that please let the Origami team know so we can ensure this tutorial is always up to date.
</aside>

## Build Tools

> Origami components are developed using the [Origami Build Tools](https://github.com/Financial-Times/origami-build-tools) (`obt`) command line interface. To work on our component, install Origami Build Tools (`obt`) on your machine:

Reviewer: ```suggestion
Origami components are developed using the [Origami Build Tools](https://github.com/Financial-Times/origami-build-tools) (`obt`) command line interface. To work on our component, install Origami Build Tools (`obt`) on your machine by running this command in your terminal:
```

### JavaScript And Sass

Origami components may include JavaScript for interactivity or [Sass](https://sass-lang.com/) for visual styles (we compile [Sass](https://sass-lang.com/) to CSS). When prompted, select that you would like the component to include JavaScript and Sass for the purposes of this tutorial. We will use the code it generates as a base to discuss building an Origami component later.

### Other Questions

This tutorial won't expand on other questions asked by the `obt init` command, like what the component description and keywords are. Please feel free to contact the Origami team if you have any questions during this step.

### Output

Now run `obt init` and answer the questions as above. You should see a directory which contains component boilerplate (commonly shared component code we can build on top of).

<figure>

> 	<img alt="" src="/assets/images/tutorial-new-component/example-init-output.png" />

Reviewer: this image is missing the GitHub team section. Should we add it as a paragraph above as well?

Author: I think it's ok to skip this, as I don't think what a GitHub team is needs more explanation outside the cli prompt. The previous section "Other Questions" doesn't list all the questions but is intended as a heads up for other questions not mentioned

	<figcaption class="o-typography-caption">
        Example output from the `obt init` command at the time of writing. It shows the questions asked and example responses, as discussed previously.
	</figcaption>
</figure>

The structure of our component now looks something like this. In the next section we will start developing our component, and in the process discuss what is in each directory.

```
o-example
├── .eslintrc.js
├── .dependabot
│   └── config.yml
├── .github
│   ├── CODEOWNERS
│   ├── ISSUE_TEMPLATE.md
│   └── workflows
│       └── [a number of .yml files]
├── .gitignore
├── .stylelintrc.js
├── README.md
├── bower.json
├── demos
│   └── src
│       ├── demo.js
│       ├── demo.mustache
│       ├── demo.scss
│       └── pa11y.mustache
├── main.js
├── main.scss
├── origami.json
├── package.json
├── src
│   ├── js
│   │   └── example.js
│   └── scss
│       ├── _brand.scss
│       └── _variables.scss
└── test
    ├── js
    │   ├── example.test.js
    │   └── helpers
    │       └── fixtures.js
    └── scss
        ├── _main.test.scss
        └── index.test.scss
```

> ## Source Control

Reviewer: I think this should/could be moved to near the end of the tutorial, into a section which is labelled "publishing your new component". I'm thinking it would help reduce the amount of knowledge the reader is exposed to upfront which is not required for them to learn about yet. What do you think @notlee ?

Author: Yeah I'd like to reduce upfront knowledge. What do you think about the "Package Management" section? I think I was preempting npm confusion but could probably split this section across the dependency install section and the publish section

Reviewer: >What do you think about the "Package Management" section?

The "Package Management" section looks good to me. 👍 

>could probably split this section across the dependency install section and the publish section

That sounds like a good idea 😄

Author: done! :D I skipped ahead

All Origami components reside in a [git](https://git-scm.com/) repository with the same name as the component. It's required that component repositories are stored remotely in one of our [github.com](https://github.com/) organisations, for example the [Financial-Times](https://github.com/Financial-Times/) organisation. There are more details about [source control in the origami specification](/spec/v1/components/#source-control).

Create a new git repository by running `git init`, and commit the boilerplate as an initial commit. For example:

<pre><code class="o-syntax-highlight--bash">git init
git add --all
git commit -m 'o-example component boilerplate'</code></pre>

You may then push this to your remote github.com repository, under the `Financial-Times` organisation or another [supported Financial Times organisation](/spec/v1/components/#source-control).

_Note: we're assuming existing git knowledge here, so please feel free to contact the Origami team if you get stuck at any point._

> You might have noticed a `.github` directory. At the time of writing, the `.github` directory itself is not part of the [file structure specification](/spec/v1/components/#files-and-folder-structure). If we deleted `.github` the component would still be an Origami component. But this directory gives us some nice features including:

Reviewer: We should think about making the `.github` directory an optional feature in the create-origami-component generator since Origami is currently the only team which uses GitHub Actions and most other teams are using CircleCI.

Author: I don't think we should do that unless we're going to provide circleci config for our users so occ etc is setup correctly. And we should check its something users actually want before creating that

- `ISSUE_TEMPLATE.md`: The contents of this file are used to provide a [template when opening a new Github issue](https://help.github.com/en/github/building-a-strong-community/about-issue-and-pull-request-templates). It helps users report bugs or provide feedback by prompting for useful information.
- `CODEOWNERS`: defines individuals or teams to automatically assign to new Github issues or pull requests ([see the Github code owners documentation](https://help.github.com/en/github/creating-cloning-and-archiving-repositories/about-code-owners)).
- `workflows/*`: the workflows directory configures a number of [Github Actions](https://github.com/features/actions) which will automate component testing and release, set useful Github labels, and more. We'll look at some of these later.

## Package Management

> Origami components use [Bower](https://bower.io/), a package manager like [NPM](https://www.npmjs.com/), to manage dependencies. The `bower.json` file lists the components dependencies, and points to the main Sass and JavaScript files of the component. One benefit of using Bower is it ensures a flat dependency tree, so two versions of the same component are not install at once.

Reviewer: ```suggestion
Origami components use [Bower](https://bower.io/), a package manager like [NPM](https://www.npmjs.com/), to manage dependencies. The `bower.json` file lists the components dependencies, and points to the main Sass and JavaScript files of the component. One benefit of using Bower is it ensures a flat dependency tree, so multiple versions of the same component are not installed at once.
```

> Although Origami components use Bower to install dependencies, developer dependencies may be installed using the [NPM](https://www.npmjs.com/) package manager, as seen in `package.json`. Rules for package management are defined in the [package management section of the specification](/spec/v1/components/#package-management).

Reviewer: ```suggestion
Although Origami components use Bower to install dependencies, developer dependencies may be installed using the [npm](https://www.npmjs.com/) package manager, as seen in `package.json`. Rules for package management are defined in the [package management section of the specification](/spec/v1/components/#package-management).
```
npm is lowercase, I won't make suggestions to replace all the uppercase versions though because there are quite a lot of them

> Although Origami components are authored using Bower, components are published to NPM so projects which use Origami may choose to use NPM over Bower ([but we still recommended Bower for now](/docs/tutorials/npm/)).

Reviewer: ```suggestion
Although Origami components are authored using Bower, components are published to NPM so projects which use Origami may choose to use NPM over Bower ([but we still recommend Bower for now](/docs/tutorials/npm/)).
```

Then we can use the `develop` (`dev`) command to start working on the component. The `dev` command creates a server for us to preview our component in the browser. And whenever we make any change to the source code the component will be rebuilt, which we will be able to see by refreshing our browser:

<pre><code class="o-syntax-highlight--bash">obt dev</code></pre>

Opening the link output by the develop command, for example `localhost:8999`, shows us the component demos (html files) and demo assets (built js and css files) in the browser.

<figure>
	<img alt="" src="/assets/images/tutorial-new-component/example-dev-output.png" />
	<figcaption class="o-typography-caption">
        Running the dev command builds our demos. We'll discuss demos more shortly.
	</figcaption>
</figure>

<figure>
	<img alt="" src="/assets/images/tutorial-new-component/example-demos-local.png" />
	<figcaption class="o-typography-caption">
        Opening the link output by the develop command, `localhost:8999` in this case, shows us the component demos (html files) and demo assets (js and css files) in the browser.
	</figcaption>
</figure>

> Clicking `demo.html` to open that demo will show a blank page. In the next section we will update this demo with markup and content for our component.

Reviewer: ```suggestion
Clicking [`demo.html`](http://localhost:8999/demo) to open that demo will show a blank page. In the next section we will update this demo with markup and content for our component.
```
I think having that be a link to the expected demo url would be helpful for the user

Author: I'd rather avoid links here mainly as the port number isn't guaranteed  but we also want users to be able to find where the demos are listed as part of this tutorial :)

## Markup

> So Origami components can be used with any backend language or framework users of Origami components copy and paste HTML from the readme `README.md` or demos in the [Origami registry](https://registry.origami.ft.com/components/) (although templates may be provided in some cases [according to the markup section of the component specification](/spec/v1/markup/#templates)). To update the markup in our component demos see the `demos` directory.

Reviewer: ```suggestion
Origami components are not built to work with any particular framework in mind. This is because product developers may choose to use any technology stack to built their application, and it’s important that they not be forced to choose a particular one in order to use Origami components. With that in mind Origami components are not written in a templating language or a framework but in plain <abbr title="Hyper Text Markup Language">HTML</abbr> and users of Origami components copy and paste the HTML from the readme `README.md` or demos in the [Origami registry](https://registry.origami.ft.com/components/). To update the markup in our component demos see the `demos` directory.
```

I've rewritten this into _hopefully_ smaller sentences and with the aim of giving more explanation about why we don't use a framework or templating language when distributing components to users.

Author: looks good, there are a couple of typos in the suggestion but I can take care of those when addressing all the feedback

> The templates for demos are written in [mustache](https://mustache.github.io/), and demos may include their own styles ([Sass](https://sass-lang.com/)) and JavaScript which is not part of the component itself. Note that demo code is not used by projects which depend on the component.

Reviewer: I don't think we've explained what demos are yet, should we explain what they are, why they are useful etc? This may be the first time a developer is making any type of components whatsoever so they may not know that components usually have live examples of how they look/work.

Author: Good point. I don't think we should go into great depth at this point but should expect  a little in the previous section. It's not the first mention of demos:

>Opening the link output by the develop command, for example `localhost:8999`, shows us the component demos (html files) and demo assets (built js and css files) in the browser.

In the demos directory, you should see an example demo `demos/src/demo.mustache` (we'll revisit the other files later). Open `demos/src/demo.mustache` in your editor and you should see something which looks like this (assuming a component name of `o-example`):

> <pre><code class="o-syntax-highlight--html">&lt;div class="o-example" data-o-component="o-example">&lt;/div></code></pre>

Reviewer: I was thinking it might be nice to have the relative path as the first line in these codeblocks to help the reader know where to go. When I got to the theming section I felt quite lost with the large codeblocks and not knowing what file they related to

That `div` element is our component markup. So we can see something in out demo, add some content within the `div`.

<pre><code class="o-syntax-highlight--diff">-&lt;div class="o-example" data-o-component="o-example">&lt;/div>
+&lt;div class="o-example" data-o-component="o-example">
+   Hello world, I am a component named o-example!
+&lt;/div></code></pre>

The `obt dev` command which we run earlier will detect that you have updated `demos/src/demo.mustache` and compile it to `demos/local/demo.html`. Now if you refresh your browser you should see the content we just added to the demo.

<figure>
	<img alt="" src="/assets/images/tutorial-new-component/hello-world-demo.png" />
	<figcaption class="o-typography-caption">
        A "hello world" component demo with no styles.
	</figcaption>
</figure>

> The `div` tag in our demo may be any HTML tag provided there is a `data-o-component` attribute. The `data-o-component` attribute identifies the root of our component and its [owned dom](/spec/v1/markup/#owned-dom). A component may act on a DOM element using JavaScript if it, or any ancestor, has a data attribute containing the component’s name. There is also a CSS class `o-example` in our demo. Origami components may only style a DOM element with CSS if it, or any ancestor, has a class which starts with the name of the component. There are more details in the [markup section of the component specification](/spec/v1/markup/) but we'll revisit this when adding CSS styles and JavaScript to our component.

Reviewer: ```suggestion
The `div` tag in our demo may be any HTML tag provided there is a `data-o-component` attribute. The `data-o-component` attribute identifies the root of our component and its [owned dom](/spec/v1/markup/#owned-dom). A component may act on a DOM element using JavaScript if the DOM element, or any ancestor, has a data attribute containing the component’s name. There is also a CSS class `o-example` in our demo. Origami components may only style a DOM element with CSS if it, or any ancestor, has a class which starts with the name of the component. There are more details in the [markup section of the component specification](/spec/v1/markup/) but we'll revisit this when adding CSS styles and JavaScript to our component.
```
